File: lib/gfa/logger.rb
#
# This class allows to output a message to the log file or STDERR and
# to keep track of the progress of a method which takes long time to complete.
#
class GFA::Logger

  # Information about the progress of a computation
  ProgressData = Struct.new(:counter, :units, :partsize,
                            :lastpart, :total, :starttime,
                            :strlen)

  # Create a Logger instance
  #
  # @param channel [#puts]
  #    where to output (default: STDERR)
  # @param prefix [String]
  #   output prefix (default: "#")
  # @param verbose_level [Integer]
  #   0: no logging; >0: the higher, the more logging
  # @return [GFA::Logger]
  def initialize(verbose_level: 1, channel: STDERR, prefix: "#")
    @progress = false
    if !verbose_level.kind_of?(Integer)
      raise "verbose_level must be an Integer"
    end
    raise "channel must provide a puts method" if !channel.respond_to?(:puts)
    @channel = channel
    @pfx = prefix
    @verbose_level = verbose_level
    @data = {}
  end

  # Output a message
  #
  # @param msg [String] message to output
  # @param min_verbose_level [Integer]
  # @return [void]
  def log(msg, min_verbose_level=1)
    @channel.puts "#@pfx #{msg}" if @verbose_level >= min_verbose_level
    return nil
  end

  # Enable output from the Logger instance
  #
  # @param part [Float]
  #  - part = 0      => output at every call of {GFA::Logger.progress_log}
  #  - 0 < part < 1  => output once per part of the total progress
  #                     (e.g. 0.001 = log every 0.1% progress)
  #  - part = 1      => output only total elapsed time
  # @return [void]
  def enable_progress(part: 0.1)
    raise "part must be in range [0..1]" if part < 0 or part > 1
    @progress = true
    @part = part
    @channel.puts "#@pfx Progress logging enabled" if @verbose_level > 0
    return nil
  end

  # Disable progress logging
  # @return [void]
  def disable_progress
    @progress = false
    @channel.puts "#@pfx Progress logging disabled" if @verbose_level > 0
    return nil
  end

  # @!macro progress_init
  #   Initialize progress logging for a computation
  #   @param symbol [Symbol] a symbol assigned to the computation
  #   @param units [String] a string with the name of the units, in plural
  #   @param total [Integer] total number of units
  #   @param initmsg [String] an optional message to output at the beginning
  # @return [void]
  def progress_init(symbol, units, total, initmsg = nil)
    return nil if !@progress or total == 0
    str = "#@pfx 0.0% #{units} processed"
    @data[symbol] = ProgressData.new(0, units, (@part*total).to_i, 1, total,
                                     Time.now, str.size)
    @channel.puts "#@pfx #{initmsg}" if initmsg
    @channel.print str if @part != 1
    return nil
  end

  # @!macro [new] progress_log
  #   Updates progress logging for a computation
  #   @!macro [new] prlog
  #     @param symbol [Symbol] the symbol assigned to the computation at
  #       init time
  #     @param keyargs [Hash] additional units to display, with their current
  #       value (e.g. segments_processed: 10000)
  #   @param progress [Integer] how many units were processed
  # @return [void]
  def progress_log(symbol, progress=1, **keyargs)
    return nil if !@progress or @part == 1
    data = @data[symbol]
    return nil if data.nil?
    data.counter += progress
    if data.counter == data.total
      progress_end(symbol)
    elsif data.partsize == 0 or
        (data.counter / data.partsize).to_i > data.lastpart
      return nil if data.partsize == 0 and @part > 0
        # this means total is very small
      data.lastpart = data.counter / data.partsize if data.partsize > 0
      done = data.counter.to_f / data.total
      t = Time.now - data.starttime
      eta = (t / done) - t
      tstr= ("Elapsed: %02dh %02dmin %02ds" % [t/3600, t/60%60, t%60])
      etastr = ("ETA: %02dh %02dmin %02ds" % [eta/3600, eta/60%60, eta%60])
      donestr = "%.1f" % (done*100)
      keystr = ""
      keyargs.each {|k,v| keystr << "; #{k}: #{v}"}
      str = "#@pfx #{donestr}% #{data.units} processed "+
              "[#{tstr}; #{etastr}#{keystr}]"
      if str.size > data.strlen
        data.strlen = str.size
        spacediff = ""
      else
        spacediff = " "*(data.strlen-str.size)
      end
      @channel.print "\r#{str}#{spacediff}"
      @channel.flush
    end
    return nil
  end

  # @!macro [new] progress_end
  #   Completes progress logging for a computation
  #   @!macro prlog
  # @return [void]
  def progress_end(symbol, **keyargs)
    return if !@progress
    data = @data[symbol]
    return if data.nil?
    t = Time.now - data.starttime
    tstr= ("Elapsed time: %02dh %02dmin %02ds" % [t/3600, t/60%60, t%60])
    quantity = @part == 1 ? data.total.to_s : "100.0%"
    keystr = ""
    keyargs.each {|k,v| keystr << "; #{k}: #{v}"}
    str = "#@pfx #{quantity} #{data.units} processed [#{tstr}#{keystr}]"
    spacediff = " "*([data.strlen - str.size,0].max)
    @channel.print "\r" if @part != 1
    @channel.puts "#{str}#{spacediff}"
    @channel.flush
    @data.delete(symbol)
    return nil
  end

end

# Progress logging related-methods for GFA class
module GFA::LoggerSupport

  # Activate logging of progress
  # @return [GFA] self
  def enable_progress_logging(part: 0.1, channel: STDERR)
    @progress = GFA::Logger.new(channel: channel)
    @progress.enable_progress(part: part)
    return self
  end

  # @!macro progress_init
  # @return [GFA] self
  def progress_log_init(symbol, units, total, initmsg = nil)
    @progress.progress_init(symbol, units, total, initmsg) if @progress
    return self
  end

  # @!macro progress_log
  # @return [GFA] self
  def progress_log(symbol, progress=1)
    @progress.progress_log(symbol, progress) if @progress
    return self
  end

  # @!macro progress_end
  # @return [GFA] self
  def progress_log_end(symbol)
    @progress.progress_end(symbol) if @progress
    return self
  end

end

File: gfapy/logger.py
import sys
import gfapy
import time

class Logger:
  """
  Output messages to the standard error or a logfile and
  keep track of the progress of long running methods.

  Parameters:
    verbose_level (int) : 0: no logging; >0: the higher, the more logging
      (default: 1); messages output using the log() method can be provided
      with a min_verbose_level, and are output only if that value is equal
      or higher than verbose_level
    channel : where to output (default: sys.stderr), it must provide a
      write() method.
    prefix (str) : output prefix (default: ```#```)

  Returns:
    gfapy.Logger
  """

  class ProgressData:
    """
    Information about the progress of a computation
    """
    def __init__(self, counter, units, partsize, lastpart, total, starttime,
                 strlen):
      self._counter = counter
      self._units = units
      self._partsize = partsize
      self._lastpart = lastpart
      self._total = total
      self._starttime = starttime
      self._strlen = strlen

  def __init__(self, verbose_level = 1, channel = sys.stderr, prefix = "#"):
    self._progress = False
    if not isinstance(verbose_level, int):
      raise gfapy.ArgumentError("verbose_level must be an Integer")
    if not(getattr(channel, "write", None) and callable(channel.write)):
      raise gfapy.TypeError("channel must provide a 'write' method")
    self._channel = channel
    self._pfx = prefix
    self._verbose_level = verbose_level
    self._data = {}

  def log(self, msg, min_verbose_level=1):
    """Output a log message to the logger output channel.

    Parameters:
      msg (str) : message to output
      min_verbose_level (int) : output the message only if the
        verbose level of the logger is at least the specified one
        (default: 1)
    """
    if self._verbose_level >= min_verbose_level:
      self._channel.write("{} {}\n".format(self._pfx, msg))

  def enable_progress(self, part = 0.1):
    """Enable output of progress of long running methods.

    Parameters
      part (float between 0 and 1) : if part = 0, output at every call of
         progress_log(); if 0 < part < 1, output once per part of the total
         progress (e.g. 0.001 = log every 0.1% progress); if part = 1, output
         only total elapsed time at the end of the computation.
    """
    if part < 0 or part > 1:
      raise gfapy.ArgumentError("part must be in range [0..1]")
    self._progress = True
    self._part = part
    if self._verbose_level > 0:
      self._channel.write("{} Progress logging enabled\n".format(self._pfx))

  def disable_progress(self):
    """Disable output of progress of long running methods."""
    self._progress = False
    if self._verbose_level > 0:
      self._channel.write("{} Progress logging disabled\n".format(self._pfx))

  def progress_init(self, symbol, units, total, initmsg = None):
    """Initialize progress logging for a long running computation.

    Parameters:
      symbol (str) : an identifier assigned to the computation
      units (str) : the name of the units of computation, in plural, for the
                    output messages
      total (int) : the total number of units of the computation
      initmsg (str) : an optional message to output at the beginning of the
                      computation
    """
    if not self._progress or total == 0:
      return
    string = "{} 0.0% {} processed".format(self._pfx, units)
    self._data[symbol] = Logger.ProgressData(0, units, int(self._part*total),
                                             1, total, time.time(), len(string))
    if initmsg:
      self._channel.write("{} {}\n".format(self._pfx, initmsg))
    if self._part != 1:
      self._channel.write(string)

  def progress_log(self, symbol, progress=1, **keyargs):
    """Updates progress of a computation.

    A logging message is output or not, depending on the part parameter
    (see the `enable_progress` method).

    Parameters:
      symbol (str) : the identifier assigned to the computation when
                    `progress_init` was called
      progress (int) : how many units of computations were completed
                       in the last interaction (default: 1)
      **keyargs (dict) : additional units of computation to display (keys),
                       together with their current progress value (values);
                       (e.g. segments_processed: 10000)
    """
    if not self._progress or self._part == 1:
      return
    data = self._data.get(symbol, None)
    if data is None:
      return
    data._counter += progress
    if data._counter == data._total:
      self.progress_end(symbol)
    elif data._partsize == 0 or \
         int(data._counter / data._partsize) > data._lastpart:
      if data._partsize == 0 and self._part > 0:
        return
      # this means total is very small
      if data._partsize > 0:
        data._lastpart = data._counter / data._partsize
      done = data._counter / data._total
      t = int(time.time() - data._starttime)
      eta = int((t / done) - t)
      tstr= ("Elapsed: {:02d}h {:02d}min {:02d}s"
             .format(t//3600, t//60%60, t%60))
      etastr = ("ETA: {:02d}h {:02d}min {:02d}s"
                .format(eta//3600, eta//60%60, eta%60))
      donestr = "{:.1f}".format(done*100)
      keystr = "".join([ "; {}: {}".format(k, v) for k,v in keyargs.items()])
      string = "{} {}% {} processed " \
               .format(self._pfx, donestr, data._units) + \
               "[{}; {}{}]" \
               .format(tstr, etastr, keystr)
      if len(string) > data._strlen:
        data._strlen = len(string)
        spacediff = ""
      else:
        spacediff = " "*(data._strlen-len(string))
      self._channel.write("\r{}{}".format(string, spacediff))
      self._channel.flush()

  def progress_end(self, symbol, **keyargs):
    """Completes progress logging for a computation.

    A message is always output. The progress is set to 100%.

    Parameters:
      symbol (str) : the identifier assigned to the computation when
                    `progress_init` was called
      **keyargs (dict) : additional units of computation to display (keys),
                       together with their final value (values);
                       (e.g. segments_processed: 100000)
    """
    if not self._progress:
      return
    data = self._data.get(symbol, None)
    if data is None:
      return
    t = int(time.time() - data._starttime)
    tstr= ("Elapsed time: {:02d}h {:02d}min {:02d}s"
           .format(t//3600, t//60%60, t%60))
    quantity = str(data._total) if self._part == 1 else "100.0%"
    keystr = "".join([ "; {}: {}".format(k, v) for k,v in keyargs.items()])
    string = "{} {} {} processed " \
             .format(self._pfx, quantity, data._units) + \
             "[{}{}]" \
             .format(tstr, keystr)
    spacediff = " " * (max(data._strlen - len(string),0))
    if self._part != 1:
      self._channel.write("\r")
    self._channel.write("{}{}\n".format(string, spacediff))
    self._channel.flush()
    self._data.pop(symbol)
